From b1d9f8ff07b7a13266cb8c37b97533f945c193b7 Mon Sep 17 00:00:00 2001
From: Volker Enderlein <volkerenderlein@hotmail.com>
Date: Fri, 29 May 2020 17:37:18 +0200
Subject: [PATCH 2/6] Fixed memory leak in SoShadowLightCache.

 fixed #388
---
 src/shadows/SoShadowGroup.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/shadows/SoShadowGroup.cpp b/src/shadows/SoShadowGroup.cpp
index dcde256e5..3d9a8cff1 100644
--- a/src/shadows/SoShadowGroup.cpp
+++ b/src/shadows/SoShadowGroup.cpp
@@ -519,7 +519,7 @@ public:
     if (this->gaussalphasceneH) this->gaussalphasceneH->unref();
     if (this->gaussalphasceneV) this->gaussalphasceneV->unref();
     if (this->depthtest) this->depthtest->unref();
-    if (this->bboxnode) this->bboxnode->ref();
+    if (this->bboxnode) this->bboxnode->unref();
     if (this->maxshadowdistance) this->maxshadowdistance->unref();
     if (this->dropoffrate) this->dropoffrate->unref();
     if (this->coscutoff) this->coscutoff->unref();
-- 
2.23.0

